### PR TITLE
Fixes & Deployment

### PR DESCRIPTION
**Pull request by adimyth**
I saw your [tweet](https://twitter.com/mihail_eric/status/1308076018050240517) today (retweeted by HugginFace) & immediately wanted to try it out. Since an application to test it out was missing, I deployed the app on Heroku and added a link. Besides, streamlit caching doesn't seem to play nicely with transformers pipeline, so removed that. 

* Removes streamlit caching, as it generates bug (refer [error.md](https://github.com/adimyth/question-answering-with-streamlit/blob/master/error.md))
* Adds poetry for package & asdf for python management
* Heroku Deployment
* Adds a gif displaying the usage
* Documentation to run locally
* Added newline at the end of files (mentioned in the PR comments)

Consider merging if you find it useful :)

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,28 +1,22 @@
...
+import wikipedia    # type: ignore
+from transformers import Pipeline, pipeline # type: ignore
 
 NUM_SENT = 10
 
 
-@st.cache
 def get_qa_pipeline() -> Pipeline:
-    qa = pipeline('question-answering')
+    qa = pipeline("question-answering", framework="tf")
```
Comment (mihail911) [inline app.py]: Can we not set this to use TF by default? Since I tested this using Pytorch, I didn't run into the Streamlit caching issue. One option is we can pass the framework (i.e. tf/pytorch) as a command line argument and then have a cached/non-cached `get_qa_pipeline` method. We can then call the correct one based on the user's inputs.

File: setup.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+echo "\
+[general]\n\
+email = \"mishraaditya6991@gmail.com\"\n\
+" > ~/.streamlit/credentials.toml
+echo "\
+[server]\n\
+headless = true\n\
+enableCORS=false\n\
+port = $PORT\n\
+" > ~/.streamlit/config.toml
```
Comment (mihail911) [inline setup.sh]: Missing a newline

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            pytest,
+            pytest-cov,
+            pytest-mock
+        ]
+
+-   repo: https://github.com/commitizen-tools/commitizen
+    rev: v1.22.3
+    hooks:
+    -   id: commitizen
+        stages: [commit-msg]
```
Comment (mihail911) [inline .pre-commit-config.yaml]: missing newline

File: .tool-versions [hunk 4]
```diff
@@ -0,0 +1,3 @@
+python 3.8.0
+poetry 1.0.10
+nodejs 14.9.0
```
Comment (mihail911) [inline .tool-versions]: missing newline

File: Procfile [hunk 5]
```diff
@@ -0,0 +1 @@
+web: sh setup.sh && streamlit run app.py
```
Comment (mihail911) [inline Procfile]: missing newline

**Comment by mihail911** [comment]
Thanks for doing this @adimyth! Looks really good. I left a few comments addressing the Streamlit caching issue. Also are you planning on continuously running a sample heroku app for this code?

File: README.md [hunk 6]
```diff
@@ -1,4 +1,44 @@
 # State-of-the-art Question Answering With Streamlit and HuggingFace
 
-This is the code accompanying [this post](https://www.mihaileric.com/posts/state-of-the-art-question-answering-streamlit-huggingface/).
+This is the code accompanying [this post](https://www.mihaileric.com/posts/state-of-the-art-question-answering-streamlit-huggingface/). You can try the app [here](https://wikipedia-transformers-qa.herokuapp.com/). Built using Streamlit and deployed on Heroku
```
Comment (mihail911) [inline README.md:3]: If the plan is to actually keep the Heroku app running, it's fine to keep this here. Otherwise if this is just a self-serve option, I would modify this to say "This setup enables you to deploy a Heroku app for the setup described in the post")

File: app.py (hunk 1 above)
Comment (adimyth) [reply app.py]: I had initially tried without adding framework parameter, but still ran into the caching issue. I'll try again and see if it works with pytorch. Will have to check why switching frameworks throws such an error. Any ideas?

Yeah, we can pass the framework as a command line argument or create a config dictionary. Either works!

File: README.md (hunk 6 above)
Comment (adimyth) [reply README.md:3]: I intend to keep it running by utilising the free dyno hours. The dyno sleeps after 30 minutes of inactivity but heroku spins up a dyno as soon as it receives a request (serve on demand). It would allow them to test it out easily (primary reason for this PR) w/o any effort. So, I think it's still better to keep the sentence as is. Besides, I haven't added steps to deploy the Heroku app in README, so it doesn't make much sense to change it to "This setup enables you to deploy a Heroku app for the setup described in the post". 

Let me know your thoughts!

File: .pre-commit-config.yaml (hunk 3 above)
Comment (adimyth) [reply .pre-commit-config.yaml]: Added newline in the new commit

File: .tool-versions (hunk 4 above)
Comment (adimyth) [reply .tool-versions]: Added newline in the new commit

File: Procfile (hunk 5 above)
Comment (adimyth) [reply Procfile]: Added newline in the new commit

File: setup.sh (hunk 2 above)
Comment (adimyth) [reply setup.sh]: Added newline in the new commit

**Comment by adimyth** [comment]
Glad that you found it useful @mihail911. Thanks for making this project in the first place! Consider tweeting along with the demo link 😉

File: app.py (hunk 1 above)
Comment (adimyth) [reply app.py]: I reran again using with pytorch 1.6, yet I face the same issue so have removed caching

File: README.md (hunk 6 above)
Comment (mihail911) [reply README.md:3]: Sounds good. As long as the app is live before this gets merged in, so people aren't confused by the wording.

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,147 @@
+# Byte-compiled / optimized / DLL files
+__pycache__/
+*.py[cod]
+*$py.class
+
+# Custom ignored
+gpt3_config.yml
```
Comment (mihail911) [inline .gitignore]: Can you remove this and other entries that seem specific to your setup? This .gitignore should ideally be specific just to this project.

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+.tool-versions
+db/results.db
```
Comment (mihail911) [inline .gitignore]: Seems unrelated?

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+# C extensions
+*.so
```
Comment (mihail911) [inline .gitignore]: Seems unrelated

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+local_settings.py
+db.sqlite3
+db.sqlite3-journal
+
+# Flask stuff:
+instance/
+.webassets-cache
+
+# Scrapy stuff:
+.scrapy
```
Comment (mihail911) [inline .gitignore]: Also seems unrelated

File: app.py [hunk 11]
```diff
@@ -1,40 +1,40 @@
...
+import streamlit as st  # type: ignore
+import wikipedia  # type: ignore
+from transformers import Pipeline, pipeline  # type: ignore
 
-NUM_SENT = 10
+from config import config
 
 
-@st.cache
 def get_qa_pipeline() -> Pipeline:
```
Comment (mihail911) [inline app.py:24]: Can we add back the `@st.cache` for the pytorch code path, since otherwise I believe it does get reloaded.

File: .gitignore (hunk 10 above)
Comment (adimyth) [reply .gitignore]: removed irrelevant entries

File: .gitignore (hunk 9 above)
Comment (adimyth) [reply .gitignore]: removed irrelevant entries

File: .gitignore (hunk 8 above)
Comment (adimyth) [reply .gitignore]: removed irrelevant entries

File: .gitignore (hunk 7 above)
Comment (adimyth) [reply .gitignore]: removed irrelevant entries

File: app.py (hunk 11 above)
Comment (adimyth) [reply app.py:24]: 1. I have added a conditional decorator, that takes in a function and returns a streamlit cached function if framework used is pytorch. 
2. Caching `answer_question` and `get_qa_pipeline()` functions provide performance boost.  It also avoids loading the transformer pipeline again as you mentioned.

Comment (mihail911) [reply app.py:24]: Looks good!

File: error.md [hunk 12]
```diff
@@ -0,0 +1,21 @@
+## Error: with streamlit caching
```
Comment (mihail911) [inline error.md]: Given your cache change above, I believe you can remove this file. Once it's removed, happy to approve.

File: setup.sh [hunk 13]
```diff
@@ -0,0 +1,11 @@
+mkdir -p ~/.streamlit/
+echo "\
+[general]\n\
+email = \"mishraaditya6991@gmail.com\"\n\
```
Comment (mihail911) [inline setup.sh]: Is providing your email here necessary?

**Review by mihail911 — APPROVED** [review]
Looks great. Thanks for doing this!
